### PR TITLE
Dashboard course list API rework

### PR DESCRIPTION
**Pull request by szabinst**
refs: MBL-16627
affects: Student, Teacher
release note: none
test plan: See ticket

This PR contains the following changes: 
- Replaced the existing dashboard course API with the one that we use for All Courses
- Dashboard folder has been rearranged. Instead of segmenting the folder by View, View Model and Model, it's segmented by (Dashboard) Sub-features. Sub-features are arranged in the same View, View Model, Model structure like other folders in the projects. 

## Screenshots

<table>
<tr><th>Dashboard folder - Before</th><th>Dashboard folder - After</th></tr>
<tr>
<td></td>
<td></td>
</tr>
</table>

## Checklist

- [ ] Follow-up e2e test ticket created or not needed
- [x] A11y checked
- [x] Tested on phone
- [x] Tested on tablet
- [x] Tested in dark mode
- [x] Tested in light mode
- [x] Approve from product or not needed

### CODE REVIEW

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Build failed, skipping coverage check</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>This pull request will not generate a release note.</td>
    </tr>
  </tbody>
</table>


#### Affected Apps: Student, Teacher

[MBL-16627](https://instructure.atlassian.net/browse/MBL-16627)

<details><summary>:x: XCTest failed: CoreTests/FilePickerViewControllerTests/testDocumentScan</summary>


    XCTAssertNotNil failed
    failed - Presented view controller should not be nil
</details>
<details><summary>:x: XCTest failed: CoreTests/FilePickerViewControllerTests/testImage</summary>


    XCTAssertNotNil failed
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 1b9895ef93c6c6eeb710b673318a1f6e36fc6380
</p>

File: Core/Core/Dashboard/K5/View/Homeroom/K5HomeroomMySubjectsView.swift [hunk 1]
```diff
@@ -64,17 +64,17 @@ public struct K5HomeroomMySubjectsView: View {
 #if DEBUG
 
 struct K5HomeroomMySubjectsView_Previews: PreviewProvider {
-    static let cards = [
+    static let courseCardListViewModel = [
```
Comment (vargaat) [inline Core/Core/Dashboard/K5/View/Homeroom/K5HomeroomMySubjectsView.swift:67]: Was this an unintentional rename width Xcode's refactor?

**Review by vargaat — CHANGES_REQUESTED** [review]

File: Core/Core/Dashboard/CourseCardList/Model/Entities/DashboardCard.swift [hunk 2]
```diff
@@ -16,6 +16,7 @@
 // along with this program.  If not, see <https://www.gnu.org/licenses/>.
 //
 
+import UIKit
```
[the remark this answers is not in the document]
Comment (szabinst) [reply Core/Core/Dashboard/CourseCardList/Model/Entities/DashboardCard.swift:19]: Removed.

File: Core/Core/Dashboard/CourseCardList/Model/DashboardCourseCardListInteractorLive.swift [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+//
+
+import Combine
+
+public class DashboardCourseCardListInteractorLive: DashboardCourseCardListInteractor {
+    // MARK: - Outputs
+
+    public let state = CurrentValueSubject<StoreState, Never>(.loading)
+    public let courseCardList: CurrentValueSubject<[DashboardCard], Never> = .init([])
+    private let courseList: CurrentValueSubject<Course, Never> = .init(.init())
```
[the remark this answers is not in the document]
Comment (szabinst) [reply Core/Core/Dashboard/CourseCardList/Model/DashboardCourseCardListInteractorLive.swift:26]: Removed.

File: Core/Core/Dashboard/CourseCardList/Model/DashboardCourseCardListInteractorLive.swift [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+    public func refresh() -> Future<Void, Never> {
+        Future { [weak self] promise in
+            guard let self else {
+                return promise(.success(()))
+            }
+            Publishers
+                .CombineLatest(
+                    self.courseCardListStore.refreshWithFuture(force: true),
+                    self.courseListStore.refreshWithFuture(force: true)
```
[the remark this answers is not in the document]
Comment (szabinst) [reply Core/Core/Dashboard/CourseCardList/Model/DashboardCourseCardListInteractorLive.swift:68]: I added a reactive extension for exhaust and updated this combineLatest with the appropriate calls.

**Comment by inst-danger** [comment]
<details><summary>Teacher Build QR Code:</summary>

</details>

**Review by vargaat — CHANGES_REQUESTED** [review]
Please check `TeacherUITests/TodoUITests/testTodos` the course's name disappeared from the todo cells.

**Comment by szabinst** [comment]
@vargaat Pushed the fix, CI is green now.

**Review by vargaat — APPROVED** [review]
QA+1

**Review by idevainstr — APPROVED** [review]
QA +1

**Pull request merged**